### PR TITLE
DID:WEB resolves to DID Resolution Result

### LINKED ISSUE #20: Decide what to do about 'proof' section and other DID Doc Metadata
Signing the DID Document itself (adding an OPTIONAL/recommended `proof` section, for example) would offer integrity protection / move towards the did docs being self-certifying.

Since the `proof` property (as well as other metadata fields such as `created` and `updated`) were removed from the DID Core spec, we need to make a decision on how to handle these in the `did:web` method.

Our options:

1. Put the metadata fields (`proof`, `created` etc) into the DID Document itself.
2. Store not the DID Document in the `.json` file on the web server, but instead embed it in a structure similar to the [DID Resolution Result](https://w3c-ccg.github.io/did-resolution/#did-resolution-result).

Example structure for option 1:

```js
{
	"@context": "https://www.w3.org/ns/did/v1",
	"id": "did:web:example:com",
	"authentication": [{
		"id": "did:web:example:com#keys-1",
		"type": "Ed25519VerificationKey2018",
		"controller": "did:web:example:com",
		"publicKeyBase58": "H3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV"
	}],
	"service": [{
		"id": "did:web:example:com#vcs",
		"type": "VerifiableCredentialService",
		"serviceEndpoint": "https://example.com/vc/"
	}],
	"proof": {
		// signature over the whole DID document would go here.
	},
	"created": "2019-03-23T06:35:22Z",
	"updated": "2023-08-10T13:40:06Z"
}
```

Example structure for option 2:

```js
{
	"@context": "https://w3id.org/did-resolution/v1",
	"didDocument": {
		"@context": "https://www.w3.org/ns/did/v1",
		"id": "did:web:example:com",
		"authentication": [{
			"id": "did:web:example:com#keys-1",
			"type": "Ed25519VerificationKey2018",
			"controller": "did:web:example:com",
			"publicKeyBase58": "H3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV"
		}],
		"service": [{
			"id": "did:web:example:com#vcs",
			"type": "VerifiableCredentialService",
			"serviceEndpoint": "https://example.com/vc/"
		}],
                "proof": {
                    // signature over the whole DID document would go here.
                }
	},
	"didDocumentMetadata": {
		"created": "2019-03-23T06:35:22Z",
		"updated": "2023-08-10T13:40:06Z"
	}
}
```

### PR DESCRIPTION
**Pull request by vitorpamplona**
Fixes #20: 
1. Changes  the nature of the `did.json` document to become a [DID Resolution Result](https://www.w3.org/TR/did-core/#did-resolution) that contains a [DID Document](https://www.w3.org/TR/did-core/#dfn-did-documents) and not the DID Document itself.
2. Establishes the presence of [DID Document Metadata](https://www.w3.org/TR/did-core/#did-document-metadata) with possibilities of adding cryptographic proofs to verify DID Documents.
3. Allows multiple [representations](https://www.w3.org/TR/did-core/#representations) (JSON, JSONLD, XML, CBOR, etc) of the DID Resolution Document for the same DID:WEB based on [DID Resolution Options](https://www.w3.org/TR/did-core/#did-resolution-options)

### Impact on current implementers

This is a backward incompatible change. 

**Current producers** must wrap the root property of their `did.json` file into a `didDocument` property. 

**DID Resolvers** should check for the presence of a `didDocument` property to indicate compliance of the producer with this new version. If the property is present, an optional `didDocumentMetadata` property can also be available and should be parsed accordingly.  

The spec's `did.json` goes from:  
```json
{
  "@context": "https://www.w3.org/ns/did/v1",
  "id": "did:web:example.com",
  "verificationMethod": [{
     "id": "did:web:example.com#controller",
     "type": "Secp256k1VerificationKey2018",
     "controller": "did:web:example.com",
     "ethereumAddress": "0xb9c5714089478a327f09197987f16f9e5d936e8a"
  }],
  "authentication": [
     "did:web:example.com#controller"
  ]
}
```

To: 

```json
{
  "@context": "https://w3id.org/did-resolution/v1",
  "didDocument": {
    "@context": "https://www.w3.org/ns/did/v1",
    "id": "did:web:example:com",
    "authentication": [{
      "id": "did:web:example:com#keys-1",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:web:example:com",
      "publicKeyBase58": "H3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV"
    }],
    "service": [{
      "id": "did:web:example:com#vcs",
      "type": "VerifiableCredentialService",
      "serviceEndpoint": "https://example.com/vc/"
    }],
  },
  "didDocumentMetadata": {
    "@context": [
      "https://w3id.org/did-resolution/v1"
      "https://w3id.org/security/suites/ed25519-2020/v1"
    ],
    "created": "2019-03-23T06:35:22Z",
    "updated": "2023-08-10T13:40:06Z",
    "proof": {
      "type": "Ed25519Signature2020",
      "created": "2022-08-21T12:54:33Z",
      "verificationMethod": "did:web:example.com:trustedSigner#WEB",
      "proofPurpose": "assertionMethod",
      "proofValue": "z3e7VcjAVz7RVb7uuH7NQzDuF9AXK3vfH65xnSJSRiLK3vTPUoCDgcE4JC1UfZhbD1wgFsWxQwdWEtrRqhbnqvyYs"
    }
  }
}
```

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -351,15 +408,22 @@ <h3>
...
           <li>
             If no path has been specified in the URL, append
             <code>/.well-known</code>.
           </li>
           <li>
-            Append <code>/did.json</code> to complete the URL.
+            Map the DID Document Option's accept property to a file extension:
+            <ol>
+              <li><code>application/did+json</code> -> <code>json</code></li>
+              <li><code>application/did+jsonld</code> -> <code>jsonld</code></li>
```
Comment (OR13) [inline index.html]:
```suggestion
              <li><code>application/did+ld+json</code> -> <code>jsonld</code></li>
```

File: index.html [hunk 2]
```diff
@@ -326,13 +269,127 @@ <h3>
...
+  "didDocumentMetadata": {
+    "created": "2019-03-23T06:35:22Z",
+    "updated": "2023-08-10T13:40:06Z",
+    "proof": {
+      "type": "Ed25519Signature2020",
+      "created": "2022-08-21T12:54:33Z",
+      "verificationMethod": "did:web:example.com:trustedSigner#WEB",
+      "proofPurpose": "assertionMethod",
+      "proofValue": "z3e7VcjAVz7RVb7uuH7NQzDuF9AXK3vfH65xnSJSRiLK3vTPUoCDgcE4JC1UfZhbD1wgFsWxQwdWEtrRqhbnqvyYs"
+    }
```
Comment (msporny) [inline index.html:351]: If you are trying to express a `proof` for the DID Document, -1 to detaching the `proof` in this way. You can include it in `didDocument`, you just need to add the appropriate context file. If we detach the `proof` in this way, we'll have to have special processing instructions just for did:web resolution which would be special just for did:web. Please move `proof` to the `didDocument` and add the appropriate `@context` definition to make it clean.

If you are trying to digitally sign the response metadata, where `proof` is is fine... but that presumes the DID Resolution context has the `Ed25519Signature2020` declaration OR you need to declare it in the didResolutionMetadata block.

File: index.html [hunk 3]
```diff
@@ -351,15 +408,22 @@ <h3>
...
+            Generate an HTTPS URL to the expected location of the DID Resolution Document
             by prepending <code>https://</code>.
           </li>
           <li>
             If no path has been specified in the URL, append
             <code>/.well-known</code>.
           </li>
           <li>
-            Append <code>/did.json</code> to complete the URL.
+            Map the DID Document Option's accept property to a file extension:
```
Comment (msporny) [inline index.html]: If we are going to do content negotiation, we should follow the HTTP standard: https://www.rfc-editor.org/rfc/rfc7231#section-5.3 or at least ask the HTTP WG what the suggested path is for Conneg these days.

**Review by msporny — CHANGES_REQUESTED** [review]
In general, +1 to this PR -- the response being a DID Resolution result makes a lot of sense.

The only concerning bits are: 1) the way the `proof` is detached from the DID Document (don't do that), and 2) the content negotiation bit.

Regarding concern 1, if you leave the `proof` as a part of the DID Document (and add an appropriate context), that would address that issue. Unless you were really trying to digitally sign the DID Document metadata, in which case, the proof is probably fine where it is?

Regarding concern 2, move the content negotiation changes into a separate PR, as that is a permathread that has been burning for a while: https://github.com/w3c-ccg/did-method-web/issues/52 ...we might just want to ask the HTTP WG what the suggested path is for Conneg these days.

File: index.html (hunk 3 above)
Comment (gribneau) [reply index.html]: > If we are going to do content negotiation, we should follow the HTTP standard

Strong +1 on following HTTP. DID core is very consistent with HTTP content negotiation.

https://www.w3.org/TR/did-core/#did-resolution-options

https://www.w3.org/TR/did-core/#did-resolution-metadata

**Comment by peacekeeper** [comment]
> -1 to detaching the `proof` in this way. You can include it in `didDocument`

I know that in VCs and Data Integrity, the assumption has always been that the `proof` property is included in the same JSON-LD object that it covers, but I would contrast that with the initial design that we had when we introduced DID metadata (based on this proposal: https://github.com/w3c/did-core/issues/65#issuecomment-597030882). We said at that time that a proof is not data about the subject, but rather metadata about either the document, or the resolution process, and that therefore it belongs into one of the two metadata buckets.

Semantically, having the `proof` inside the DID document would be a bit like having the `proof` inside the `credentialSubject` object, no?

Structurally, having the `proof` outside the DID document would make it easier to cleanly distinguish between proofs that are added by the DID controller or the DID method (DID document metadata), and proofs that are added by a resolver (DID resolution metadata).

From a pure RDF/JSON-LD perspective, I think there shouldn't be a strict necessity to have the `proof` object directly under the object it applies to, or?

**Review by peacekeeper — CHANGES_REQUESTED** [review]
I don't really have a strong opinion whether or not `did:web` should be changed, but if the decision is to support metadata, then this needs to be better aligned with the [DID Resolution specification](https://w3c-ccg.github.io/did-resolution/), e.g.:

- Use of different names for the same thing "DID Resolution Document" vs. "DID Resolution Result".
- The DID Resolution specification already defines an HTTP(S) mapping of the DID Resolution process (e.g. the use of the Accept: header), so there should be no need to re-define these things here.
- This PR seems to introduce a strange new JSON-LD context `https://www.w3.org/ns/didMetadata/v1`, even though `https://w3id.org/did-resolution/v1` already defines terms for metadata.

**Comment by vitorpamplona** [comment]
@peacekeeper Thanks Markus! I fixed them all. 

I think the text could use a lot of improvements outside the context of this PR to be better aligned with the DID Resolution specification. I am just trying to not change everything. :) 

For instance, I think we need to force DNSSec just like we force HTTPS. But that is for another day.

**Review by mprorock — CHANGES_REQUESTED** [review]
Not really interested in breaking changes right now

**Comment by vitorpamplona** [comment]
> Not really interested in breaking changes right now

Hi Mike, any suggestion on how to add the DID Document Metadata with backwards compatibility?

**Comment by letmaik** [comment]
Doesn't this prevent the actual resolver from adding its own metadata?

**Comment by peacekeeper** [comment]
> Doesn't this prevent the actual resolver from adding its own metadata?

There are two "buckets" of metadata. [DID document metadata](https://www.w3.org/TR/did-core/#did-document-metadata) and [DID resolution metadata](https://www.w3.org/TR/did-core/#did-resolution-metadata). Metadata about the resolution process itself should only go into the second bucket.

**Comment by OR13** [comment]
I suggest we close this PR, and perhaps tackle some of the concepts addressed here using additional headers.

**Comment by OR13** [comment]
Closing this PR (it does not seem like it can reach consensus).

I am open to discussing on issues, feel free to file, and link here.

**Pull request closed without merging**